docs2/site/docs/getting-started/custom-scalars.md

```diff
@@ -183,105 +213,303 @@ public class ExampleSchema : Schema
+public override object ParseValue(object value)
 {
-    public IValue Convert(object value, IGraphType type)
+    if (value == null)
+        return null;
+
+    if (value is string vector3InputString)
     {
-        return new Vector3Value((Vector3)value);
+        try
+        {
+            var vector3Parts = vector3InputString.Split(',');
+            var x = float.Parse(vector3Parts[0]);
+            var y = float.Parse(vector3Parts[1]);
+            var z = float.Parse(vector3Parts[2]);
+            return new Vector3(x, y, z);
+        }
+        catch
+        {
+            throw new FormatException($"Failed to parse {nameof(Vector3)} from input '{vector3InputString}'. Input should be a string of three comma-separated floats in X Y Z order, ex. 1.0,2.0,3.0");
+        }
     }
 
-    public bool Matches(object value, IGraphType type)
+    if (value is IDictionary<string, object> dictionary)
     {
-        return value is Vector3;
+        try
+        {
+            var x = Convert.ToSingle(dictionary["x"]);
+            var y = Convert.ToSingle(dictionary["y"]);
+            var z = Convert.ToSingle(dictionary["z"]);
+            if (dictionary.Count > 3)
+                return ThrowValueConversionError(value);
+            return new Vector3(x, y, z);
+        }
+        catch
+        {
+            throw new FormatException($"Failed to parse {nameof(Vector3)} from object. Input should be an object of three floats named X Y and Z");
+        }
     }
+
+    return ThrowValueConversionError(value);
 }
 ```
 
-Register `Vector3AstValueConverter` with the schema. Don't conflate `ValueConverter.Register` used in step 4 with `Schema.RegisterValueConverter` - the latter is used for conversions to AST value nodes.
+2. Change `Serialize` to return structured data.
 
 ```csharp
-// In ExampleSchema
-public ExampleSchema(IServiceProvider provider)
-    : base(provider)
-{        
-    ValueConverter.Register<string, Vector3>(ParseVector3);
-    RegisterValueConverter(new Vector3AstValueConverter());
-
-    // Other schema assignments...
+// In Vector3Type
+
+public override object Serialize(object value)
+{
+    if (value == null)
+        return null;
+
+    if (value is Vector3 vector3)
+        return vector3;
+
+    return ThrowSerializationError(value);
 }
-``` 
+```
+
+3. Change `ToAST` to return an AST literal that represents the data.
 
-Update the implementation of `Vector3Type.ParseLiteral`:
+Since `Serialize` no longer returns a type that can be converted to an AST node, it is
+necessary to override this method.
 
 ```csharp
 // In Vector3Type
-public override object ParseLiteral(IValue value)
+
+public override IValue ToAST(object value)
 {
-    // new test
-    if (value is Vector3Value vector3Value)
-        return ParseValue(vector3Value.Value);
+    if (value == null)
+        return new NullValue();
+
+    if (value is Vector3 vector3)
+    {
+        return new StringValue($"{vector3.X},{vector3.Y},{vector3.Z}");
```

Comment:
AST should be a string or a structure (object)?

Reply:
Must be a string, as it needs to be able to be parsed by ParseLiteral.  From what I understand of the graphql spec, and based on the comments in the spec repo, scalar representations within graphql language cannot be structured.  Only the representations as variables or as serialized may be structured.

In my mind, `ParseLiteral` and `ToAST` are inverse functions of each other, and `ParseValue` and `Serialize` are inverse functions of each other.

Reply:
I think technically there is nothing preventing us to parse `ObjectValue` as literal. What is the difference at least for us parsing input object or scalar from this structured data? Both will work. Right?

Reply:
I think technically there is nothing preventing us to parse `ObjectValue` as literal. What is the difference at least for us parsing input object or scalar from this structured data? Both will work. Right?

Reply:
Try one more test supplying custom scalar as structured (object) literal.

Reply:
Have not tried it yet, and I think it would work, except pretty sure it fails validation (as it should). It would be interesting to try though.

Reply:
Yes, interesting.

Reply:
It works!!!  Tests added.

Reply:
👍